File: JULY-2025/09-07-2025/p3439.rb
class Numeric
    def max(v) = (self > v ? self : v)
end

class Solver
    class Mtg
        attr_reader :s, :e

        def initialize(s, e)
            @s, @e = s, e
        end

        def dur = (e - s)
    end

    attr_reader :ge, :mtgs

    def initialize(ge, st, et)
        @ge, @mtgs = ge, st.zip(et).collect {|(s, e)| Mtg.new(s, e) }
    end

    def solve(k)
        b, wsum = 0, 0
        (mtgs + [Mtg.new(ge, ge)]).each_with_index.inject(0) {|res, (mtg, idx)|
            wsum += mtg.dur - (idx <= k ? 0 : mtgs[idx - k - 1].dur)
            b = mtgs[idx - k - 1].e if idx > k
            res.max(mtg.e - b - wsum)
        }
    end
end

def max_free_time(event_time, k, start_time, end_time) = Solver.new(event_time, start_time, end_time).solve(k)

File: JULY-2025/09-07-2025/p3439.php
class Solution {

    /**
     * @param Integer $eventTime
     * @param Integer $k
     * @param Integer[] $startTime
     * @param Integer[] $endTime
     * @return Integer
     */
    function maxFreeTime($eventTime, $k, $startTime, $endTime) {
        $count = count($startTime);
    $max = 0;
    $prefixSum = array_fill(0, $count + 1, 0);

    for ($i = 0; $i < $count; $i++) {
        $prefixSum[$i + 1] = $prefixSum[$i] + ($endTime[$i] - $startTime[$i]);
    }

    for ($i = $k - 1; $i < $count; $i++) {
        $occupied = $prefixSum[$i + 1] - $prefixSum[$i - $k + 1];
        $windowEnd = ($i == $count - 1) ? $eventTime : $startTime[$i + 1];
        $windowStart = ($i == $k - 1) ? 0 : $endTime[$i - $k];
        $max = max($max, $windowEnd - $windowStart - $occupied);
    }

    return $max;
    }
}